Making ops c10-full: Generator arguments

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49259 Remove `use_c10_dispatcher: full` lines
* #49251 Remove generated_unboxing_wrappers and setManuallyBoxedKernel
* #49220 Remove .impl_UNBOXED() and functionalities associated with it
* #49164 Remove codegen logic to support non-c10-full ops
* #49138 [wip] Making ops c10-full: list of optional tensors
* #49150 Making ops c10-full: ConstQuantizerPtr
* #49146 Making ops c10-full: Storage arguments
* #49088 Making ops c10-full: optional lists
* #49083 Making ops c10 full: optional out arguments
* **#49013 Making ops c10-full: Generator arguments**
* #49012 Making ops c10 full: out overloads with default arguments
* #49008 Making ops c10-full: Dimname arguments
* #49007 Making ops c10-full: ops blocked by manual registrations
* #49006 Making ops c10 full: ops returning multiple out arguments
* #48912 Make out ops c10-full (with hacky-wrapper)

I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

smessmer committed:
Making ops c10-full: Generator arguments

I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

bhosmer reviewed (commented):
I don't think it's plausible to stop dispatching on Generator for unboxed, is it? That would necessitate a redesign of the current RNG approach AIUI. 

Agree that we should teach DispatchKeyExtractor about Generator and run some perf tests. But I don't imagine Generator scanning per se will affect perf much - I'd think the change would be dominated by the move from generated unboxing wrappers to the boxed protocol, no? (And as such, should we have some comparison numbers further up the stack for other ops that have made the same transition?)

bhosmer commented:
onnx errors look real btw

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer commented:
I found an old issue about this: https://github.com/pytorch/pytorch/issues/36733
This issue is long standing and this PR shouldn't change anything, it doesn't work before and it doesn't work after. I think we can go ahead and land this.

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]

smessmer committed:
Update on "Making ops c10-full: Generator arguments"


I don't know why this works. I know, this is never a good way to start a PR description :P
I know that Generator is a dispatch relevant argument when called from an unboxed API and is ignored
for dispatch purposes when called from a boxed API. This should break something, but maybe we don't
have test cases for that.

We likely need to align the unboxed and boxed dispatch behavior before landing this.
The best solution would be to make Generator not dispatch relevant in unboxing. But that might be a bigger change.
An acceptable solution could be to make Generator dispatch relevant in boxing, but that needs perf measurements.

This PR needs further discussion.

Differential Revision: [D25394998](https://our.internmc.facebook.com/intern/diff/D25394998/)

[ghstack-poisoned]
